lib/Driver/ToolChains.cpp

```diff
@@ -1078,6 +1079,74 @@ ToolChain::constructInvocation(const LinkJobAction &job,
   llvm_unreachable("linking not implemented for this toolchain");
 }
 
+ToolChain::InvocationInfo
+ToolChain::constructInvocation(const ArchiveJobAction &job,
+                               const JobContext &context) const {
+   assert(context.Output.getPrimaryOutputType() == file_types::TY_Image &&
+         "Invalid linker output type.");
+  assert(job.getKind() == LinkKind::StaticLibrary);
+
+  ArgStringList Arguments;
+
+  // Configure the toolchain.
+  bool isLLVMAR = false;
+  const char *AR = nullptr;
+  if (const Arg *A = context.Args.getLastArg(options::OPT_tools_directory)) {
+    StringRef toolchainPath(A->getValue());
+
+    // If there is an llvm-ar in the toolchain folder, use that instead.
+    if (auto toolchainAR =
+            llvm::sys::findProgramByName("llvm-ar", {toolchainPath})) {
+      AR = context.Args.MakeArgString(toolchainAR.get());
+      isLLVMAR = true;
+    }
+
+  }
+  if (AR == nullptr) {
+    if (auto pathAR = llvm::sys::findProgramByName("llvm-ar", None)) {
```

Comment:
I'm not sure we actually want to prefer `llvm-ar` like this, and I'm not sure whether this belongs in the default ToolChain either (rather than GenericUnix). On Apple platforms, `libtool -static` is what Xcode uses to create static libraries, not `ar`.

@cooperp, what's the current recommendation on this? `ar` or `libtool` for static libraries on Apple platforms?

Reply:
I believe it's libtool.  Otherwise you tend to have to call ranlib after ar, while the man page for libtool says it can handle both for you: "Libtool with -static is intended to  replace  ar(5)  and  ranlib."

Reply:
@compnerd, is what’s here what we want on both GenericUnix and Windows, or should the behaviour differ?

Reply:
@cooperp - thats the trick wrt `ar crs` - it creates the index as part of the invocation.  With `llvm-ar` that also means that we can invoke it as `ar --format bsd crs ...` which will create the BSD style archive with the index even on a non-BSD host.

Reply:
Xcode doesn't ship with `llvm-ar`. Please stick to `libtool`. (In general, please stick to "what Xcode does when you make a static library from a template".)

Reply:
Sure, but it does ship with `ar` from cctools, which also supports the `crs` mode.  The problem is that you lose compatibility with AT&T Unix, which I believe didn't support the `s` option.

Reply:
Hm. I'll let @cooperp decide what's right here. We already aren't getting code sharing because the Darwin toolchain deliberately isn't a GenericUnix toolchain, on account of Apple <del>doing weird things</del> having more control over the tools it ships.

Reply:
@cooperp, what do you think is best here?